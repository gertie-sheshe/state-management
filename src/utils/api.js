const API_URL = "http://localhost:3000";
const HEADERS = {
  Accept: "application/json",
  "Content-Type": "application/json",
};

export const apiRequest = async (url, options) => {
  const result = await fetch(url, options);
  const data = await result.json();

  return data;
};

export const getUsers = () => {
  return apiRequest(`${API_URL}/users`);
};

export const getTodos = () => {
  return apiRequest(`${API_URL}/todos`);
};

export const getUserTodos = (userId) => {
  return apiRequest(`${API_URL}/users/${userId}/todos`);
};

export const updateTodo = (todoId, data) => {
  return apiRequest(`${API_URL}/todos/${todoId}`, {
    method: "PUT",
    headers: HEADERS,
    body: JSON.stringify(data),
  });
};

export const createTodo = (data, userId) => {
  return apiRequest(`${API_URL}/users/${userId}/todos`, {
    method: "POST",
    headers: HEADERS,
    body: JSON.stringify(data),
  });
};
