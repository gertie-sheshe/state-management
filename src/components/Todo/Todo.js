import React, { useState, useEffect, useRef } from "react";
import { useMutation, useQueryClient } from "react-query";
import { updateTodo } from "../../utils/api";

function Todo({ todo }) {
  const [isChecked, setIsChecked] = useState(todo.completed);
  const [isDisabled, setIsDisabled] = useState(false);
  const isInitialMount = useRef(true);
  const queryClient = useQueryClient();

  const updateTodoMutation = useMutation(
    (options) => {
      const { data, id } = options;
      updateTodo(data, id);
    },
    {
      onSuccess: () => {
        queryClient.invalidateQueries("userTodos", todo.userId);
        setIsDisabled(!isDisabled);
      },
    }
  );

  useEffect(() => {
    if (isInitialMount.current) {
      isInitialMount.current = false;
    } else {
      const updatedTodo = { ...todo, completed: isChecked };
      updateTodoMutation.mutate({ data: updatedTodo, id: todo.id });
    }
  }, [isChecked]);

  const handleOnChange = () => {
    setIsChecked(!isChecked);
    setIsDisabled(!isDisabled);
  };

  return (
    <li key={todo.id}>
      <input
        disabled={isDisabled}
        onChange={handleOnChange}
        id={todo.id}
        type="checkbox"
        checked={isChecked}
      />
      <label htmlFor={todo.id}>{todo.title}</label>
    </li>
  );
}

export default Todo;
