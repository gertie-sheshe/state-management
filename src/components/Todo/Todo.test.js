import { render } from "../../tests/utils";
import Todo from "./Todo";

describe("Todo", () => {
  it("matches snapshot", () => {
    const { asFragment } = render(<Todo todo={{ id: 1, title: "test" }} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
