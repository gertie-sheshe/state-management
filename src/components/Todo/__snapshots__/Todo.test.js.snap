// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo matches snapshot 1`] = `
<DocumentFragment>
  <li>
    <input
      id="1"
      type="checkbox"
    />
    <label
      for="1"
    >
      test
    </label>
  </li>
</DocumentFragment>
`;
